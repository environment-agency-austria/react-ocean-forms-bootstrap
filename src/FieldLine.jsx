/**
 * Copyright (c) 2018-present, Umweltbundesamt GmbH
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import React from 'react';
import PropTypes from 'prop-types';
import FontAwesomeIcon from '@fortawesome/react-fontawesome';
import faSpinner from '@fortawesome/fontawesome-free-solid/faSpinner';
import faExclamationCircle from '@fortawesome/fontawesome-free-solid/faExclamationCircle';
import {
  FormGroup,
  Label,
  Col,
  Button,
  Alert,
  InputGroup,
  InputGroupAddon,
} from 'reactstrap';
import { fieldMetaShape, fieldShape, validators as defaultValidators } from 'react-ocean-forms';

import { BaseFieldError } from './FieldError';

/**
 * Component for displaying bootstrap
 * form groups with any children
 */
class FieldLine extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      infoVisible: false,
    };

    this.toggleInfo = this.toggleInfo.bind(this);
  }

  /**
   * Toggles the visibility of the info alert
   */
  toggleInfo() {
    this.setState(prevState => ({
      infoVisible: !prevState.infoVisible,
    }));
  }

  /**
   * Creates a * marker for the label if the field
   * contains a default required validator
   */
  createRequiredMarker() {
    const {
      validators,
      meta,
    } = this.props;

    if (meta.plaintext) {
      // Hide the required star in plaintext mode
      return null;
    }

    if (Array.isArray(validators) && validators.includes(defaultValidators.required)) {
      const requiredTitle = meta.stringFormatter('ojs_field_required');
      return <span className="field-required" title={requiredTitle}> *</span>;
    }

    return null;
  }

  /**
   * Creates a info button addon if the field
   * has a info property provided
   */
  createInfoAddonButton() {
    const {
      info,
    } = this.props;

    if (info) {
      return (
        <InputGroupAddon addonType="append">
          <Button onClick={this.toggleInfo} outline>
            <FontAwesomeIcon icon={faExclamationCircle} />
          </Button>
        </InputGroupAddon>
      );
    }

    return null;
  }

  createAddon(item, type) {
    if (!item) return null;

    const { meta } = this.props;

    let child = null;
    if (typeof item === 'function') {
      child = item();
    } else {
      child = meta.stringFormatter(item);
    }

    return (
      <InputGroupAddon addonType={type}>
        {child}
      </InputGroupAddon>
    );
  }

  createPrefix() {
    const { prefix } = this.props;
    return this.createAddon(prefix, 'prepend');
  }

  createSuffix() {
    const { suffix } = this.props;
    return this.createAddon(suffix, 'append');
  }

  /**
   * Creates a info alert if the field
   * has a info property provided
   */
  createInfoAlert() {
    const {
      info,
      meta,
    } = this.props;

    const {
      infoVisible,
    } = this.state;

    if (info) {
      const infoString = meta.stringFormatter(info);
      return (
        <Alert
          color="success"
          className="mt-2"
          isOpen={infoVisible}
          toggle={this.toggleInfo}
        >
          {infoString}
        </Alert>
      );
    }

    return null;
  }

  render() {
    const {
      field,
      meta,
      label,
      children,
    } = this.props;

    // Generate optional elements
    const requiredMarker = this.createRequiredMarker();
    const infoButton = this.createInfoAddonButton();
    const infoAlert = this.createInfoAlert();

    const invalidAlert = meta.valid
      ? null
      : <FontAwesomeIcon icon={faExclamationCircle} className="ico-invalid mr-2" />;

    const validatingSpinner = meta.isValidating
      ? <FontAwesomeIcon icon={faSpinner} spin className="ico-loading mr-2" />
      : null;

    const groupClass = meta.valid ? '' : 'is-invalid';
    const labelString = meta.stringFormatter(label);

    const prefixItem = this.createPrefix();
    const suffixItem = this.createSuffix();

    return (
      <FormGroup row className={groupClass}>
        <Label sm="3" for={field.id} className="text-right">
          {invalidAlert}
          {validatingSpinner}
          {labelString}
          {requiredMarker}
        </Label>
        <Col sm="9">
          <InputGroup>
            {prefixItem}
            {children}
            {suffixItem}
            {infoButton}
            <BaseFieldError
              id={`${field.id}_errors`}
              invalid={!meta.valid}
              error={meta.error}
              context={{ stringFormatter: meta.stringFormatter }}
            />
          </InputGroup>

          {infoAlert}
        </Col>
      </FormGroup>
    );
  }
}

FieldLine.defaultProps = {
  info: undefined,
  validators: undefined,
  prefix: undefined,
  suffix: undefined,
};

FieldLine.propTypes = {
  label: PropTypes.string.isRequired,
  info: PropTypes.string,
  meta: fieldMetaShape.isRequired,
  field: fieldShape.isRequired,
  children: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.node),
    PropTypes.node,
  ]).isRequired,
  validators: PropTypes.arrayOf(PropTypes.func),
  prefix: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.func,
  ]),
  suffix: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.func,
  ]),
};

export default FieldLine;
