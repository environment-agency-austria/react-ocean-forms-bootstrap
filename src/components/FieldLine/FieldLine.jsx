/**
 * Copyright (c) 2018-present, Umweltbundesamt GmbH
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
import React from 'react';
import PropTypes from 'prop-types';
import { Label, Col, InputGroup } from 'reactstrap';
import { FormText } from 'react-ocean-forms';
import { fieldMetaShape, fieldShape } from 'react-ocean-forms-legacy';

import { FieldRow } from './FieldRow';
import { RequiredMarker } from './RequiredMarker';
import { InvalidAlert } from './InvalidAlert';
import { ValidatingSpinner } from './ValidatingSpinner';
import { FieldLineAddon } from './FieldLineAddon';
import { InfoAddonButton } from './InfoAddonButton';
import { InfoAlert } from './InfoAlert';
import FieldError from './FieldError';

/**
 * Component for displaying bootstrap
 * form groups with any children
 */
class FieldLine extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      infoVisible: false,
    };

    this.toggleInfo = this.toggleInfo.bind(this);
  }

  /**
   * Toggles the visibility of the info alert
   */
  toggleInfo() {
    this.setState(prevState => ({
      infoVisible: !prevState.infoVisible,
    }));
  }

  render() {
    const {
      field,
      meta,
      label,
      children,
      validators,
      prefix,
      suffix,
      info,
      labelSize,
      inputSize,
      labelClass,
    } = this.props;

    const { infoVisible } = this.state;

    return (
      <FieldRow meta={meta}>
        <Label sm={labelSize} for={field.id} className={labelClass}>
          <InvalidAlert valid={meta.valid} />
          <ValidatingSpinner isValidating={meta.isValidating} />
          <FormText text={label} />
          <RequiredMarker meta={meta} validators={validators} />
        </Label>
        <Col sm={inputSize}>
          <InputGroup>
            <FieldLineAddon plaintext={meta.plaintext} type="prepend" content={prefix} />
            {children}
            <FieldLineAddon plaintext={meta.plaintext} type="append" content={suffix} />
            <InfoAddonButton info={info} plaintext={meta.plaintext} onClick={this.toggleInfo} />
            <FieldError
              id={`${field.id}_errors`}
              invalid={!meta.valid}
              error={meta.error}
            />
          </InputGroup>
          <InfoAlert
            visible={infoVisible}
            info={info}
            plaintext={meta.plaintext}
            onClose={this.toggleInfo}
          />
        </Col>
      </FieldRow>
    );
  }
}

FieldLine.displayName = 'FieldLine';

FieldLine.defaultProps = {
  info: undefined,
  validators: undefined,
  prefix: undefined,
  suffix: undefined,
  labelSize: 3,
  inputSize: 9,
  labelClass: 'text-right',
};

FieldLine.propTypes = {
  label: PropTypes.string.isRequired,
  info: PropTypes.string,
  meta: fieldMetaShape.isRequired,
  field: fieldShape.isRequired,
  children: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.node),
    PropTypes.node,
  ]).isRequired,
  validators: PropTypes.arrayOf(PropTypes.func),
  prefix: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.node,
    PropTypes.func,
  ]),
  suffix: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.node,
    PropTypes.func,
  ]),
  labelSize: PropTypes.string,
  inputSize: PropTypes.string,
  labelClass: PropTypes.string,
};

export default FieldLine;
