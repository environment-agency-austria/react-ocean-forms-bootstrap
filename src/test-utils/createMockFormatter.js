export default () => jest.fn().mockImplementation(id => id);
