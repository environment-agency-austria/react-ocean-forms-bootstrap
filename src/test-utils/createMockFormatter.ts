export const createMockFormatter = (): jest.Mock => jest.fn().mockImplementation((id: string) => id);
