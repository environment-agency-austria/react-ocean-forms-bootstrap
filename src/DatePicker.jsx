/**
 * Copyright (c) 2018-present, Umweltbundesamt GmbH
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import React from 'react';
import PropTypes from 'prop-types';
import Datetime from 'react-datetime';
import moment from 'moment';
import { Input as StrapInput } from 'reactstrap';
import { fieldMetaShape, fieldShape } from 'react-ocean-forms';

import FieldLine from './FieldLine';

/**
 * Component for displaying datetime
 */
class DatePicker extends React.Component {
  constructor(props) {
    super(props);

    this.handleChange = this.handleChange.bind(this);
  }

  handleChange(value) {
    if (value instanceof moment) {
      const { field } = this.props;
      field.onChange({
        target: {
          name: field.name,
          value: value.format(),
        },
      });
    }
  }

  render() {
    const {
      field,
      dateFormat,
      timeFormat,
      placeholder,
      meta,
    } = this.props;

    const formattedValue = moment(field.value);
    const inputProps = {
      disabled: field.disabled,
    };

    // Support for plaintext display
    if (meta.plaintext) {
      let displayValue = '';
      let parsedFormat = '';

      if (typeof dateFormat === 'string') {
        parsedFormat = dateFormat;
      } else if (dateFormat === true) {
        parsedFormat = 'L';
      }

      if (typeof timeFormat === 'string') {
        parsedFormat = `${parsedFormat} ${timeFormat}`.trim();
      } else if (timeFormat === true) {
        parsedFormat = `${parsedFormat} LT`.trim();
      }

      displayValue = formattedValue.format(parsedFormat === '' ? undefined : parsedFormat);

      return (
        <FieldLine {...this.props}>
          <StrapInput {...field} plaintext>{displayValue}</StrapInput>
        </FieldLine>
      );
    }

    return (
      <FieldLine {...this.props}>
        <Datetime
          id={field.id}
          value={formattedValue}
          placeholder={placeholder}
          onChange={this.handleChange}
          dateFormat={dateFormat}
          timeFormat={timeFormat}
          inputProps={inputProps}
          closeOnSelect
        />
      </FieldLine>
    );
  }
}

DatePicker.displayName = 'DatePicker';

DatePicker.defaultProps = {
  info: undefined,
  placeholder: 'ojs_select_placeholder',
  dateFormat: null,
  timeFormat: null,
};

DatePicker.propTypes = {
  label: PropTypes.string.isRequired,
  info: PropTypes.string,
  meta: fieldMetaShape.isRequired,
  field: fieldShape.isRequired,
  placeholder: PropTypes.string,
  dateFormat: PropTypes.oneOfType([PropTypes.string, PropTypes.bool]),
  timeFormat: PropTypes.oneOfType([PropTypes.string, PropTypes.bool]),
};

export default DatePicker;
