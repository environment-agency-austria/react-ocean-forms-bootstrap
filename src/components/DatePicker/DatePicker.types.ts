import { IFieldComponentProps } from 'react-ocean-forms';
import { DefaultizedFieldLineProps } from '../FieldLine';

/**
 * Props for the DatePicker component
 */
export interface IDatePickerProps extends IFieldComponentProps, DefaultizedFieldLineProps {
  dateFormat?: string | boolean;
  timeFormat?: string | boolean;
  inputFormat?: string | [string];
}
