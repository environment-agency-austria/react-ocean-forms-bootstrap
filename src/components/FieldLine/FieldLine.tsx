/**
 * Copyright (c) 2018-present, Umweltbundesamt GmbH
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
import * as React from 'react';

import { FormText, PropsOf } from 'react-ocean-forms';
import { Col, InputGroup, Label } from 'reactstrap';

import { FieldError } from './FieldError';
import { FieldLineAddon } from './FieldLineAddon';
import { FieldRow } from './FieldRow';
import { InfoAddonButton } from './InfoAddonButton';
import { InfoAlert } from './InfoAlert';
import { InvalidAlert } from './InvalidAlert';
import { RequiredMarker } from './RequiredMarker';
import { ValidatingSpinner } from './ValidatingSpinner';

import { IFieldLineProps } from './FieldLine.types';

interface IFieldLineState {
  infoVisible: boolean;
}

/**
 * Component for displaying bootstrap
 * form groups with any children
 */
export class FieldLine extends React.Component<IFieldLineProps, IFieldLineState> {
  public static displayName: string = 'FieldLine';

  public static defaultProps = {
    labelSize: '3',
    inputSize: '9',
    labelClass: 'text-right',
  };

  constructor(props: IFieldLineProps) {
    super(props);

    this.state = {
      infoVisible: false,
    };
  }

  /**
   * Toggles the visibility of the info alert
   */
  private toggleInfo = (): void => {
    this.setState(prevState => ({
      infoVisible: !prevState.infoVisible,
    }));
  }

  public render(): JSX.Element {
    const {
      field,
      meta,
      label,
      children,
      prefix,
      suffix,
      info,
      labelSize,
      inputSize,
      labelClass,
    } = this.props;

    const { infoVisible } = this.state;

    return (
      <FieldRow meta={meta}>
        <Label sm={labelSize} for={field.id} className={labelClass}>
          <InvalidAlert valid={meta.valid} />
          <ValidatingSpinner isValidating={meta.isValidating} />
          <FormText text={label} />
          <RequiredMarker meta={meta} />
        </Label>
        <Col sm={inputSize}>
          <InputGroup>
            <FieldLineAddon plaintext={meta.plaintext} type="prepend" content={prefix} />
            {children}
            <FieldLineAddon plaintext={meta.plaintext} type="append" content={suffix} />
            <InfoAddonButton info={info} plaintext={meta.plaintext} onClick={this.toggleInfo} />
            <FieldError
              id={`${field.id}_errors`}
              invalid={!meta.valid}
              error={meta.error}
            />
          </InputGroup>
          <InfoAlert
            visible={infoVisible}
            info={info}
            plaintext={meta.plaintext}
            onClose={this.toggleInfo}
          />
        </Col>
      </FieldRow>
    );
  }
}

export type DefaultizedFieldLineProps = JSX.LibraryManagedAttributes<typeof FieldLine, PropsOf<typeof FieldLine>>;
