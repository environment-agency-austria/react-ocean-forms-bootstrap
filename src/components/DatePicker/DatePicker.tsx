/**
 * Copyright (c) 2018-present, Umweltbundesamt GmbH
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import * as React from 'react';

import moment from 'moment';

import { default as Datetime } from 'react-datetime';
import { withField } from 'react-ocean-forms';
import { Input as StrapInput } from 'reactstrap';

import { FieldLine } from '../FieldLine';
import { IDatePickerProps } from './DatePicker.types';

/**
 * Component for displaying datetime
 */
export class BaseDatePicker extends React.Component<IDatePickerProps> {
  public static displayName: string = 'DatePicker';

  private handleBlur = (value: moment.Moment | string): void => {
    const { field, inputFormat } = this.props;

    const parsed = moment(value, inputFormat);

    const formatted = (parsed.isValid()) ? parsed : value;

    if (moment.isMoment(formatted)) {
      field.onChange({
        target: {
          value: formatted.format(),
        },
      });
    } else if (value === '') {
      field.onChange({
        target: {
          value,
        },
      });
    } else {
      field.onChange({
        target: {
          value: '',
        },
      });
    }
  }

  private handleChange = (value: moment.Moment | string): void => {
    const { field } = this.props;

    if (moment.isMoment(value)) {
      field.onChange({
        target: {
          value: value.format(),
        },
      });
    } else if (value === '') {
      field.onChange({
        target: {
          value,
        },
      });
    }
  }

  /**
   * Mimicks the behaviour of react-datetime to display
   * the same value in plaintext mode
   */
  private getDisplayValue = (value: moment.Moment): string => {
    const { dateFormat, timeFormat } = this.props;

    let parsedFormat = '';

    if (!value.isValid()) {
      return '';
    }

    if (typeof dateFormat === 'string') {
      parsedFormat = dateFormat;
    } else if (dateFormat === true) {
      parsedFormat = 'L';
    }

    if (typeof timeFormat === 'string') {
      parsedFormat = `${parsedFormat} ${timeFormat}`.trim();
    } else if (timeFormat === true) {
      parsedFormat = `${parsedFormat} LT`.trim();
    }

    return value.format(parsedFormat === '' ? undefined : parsedFormat);
  }

  public render(): JSX.Element {
    const {
      field,
      dateFormat,
      timeFormat,
      meta,
    } = this.props;

    const fieldValue = field.value;
    if (typeof fieldValue !== 'string' && typeof fieldValue !== 'number'
        && !moment.isMoment(fieldValue) && fieldValue !== undefined) {
      throw new Error(
        'Incompatible field value supplied for input component '
        + `${field.id}. Only values with type string, number or undefined are allowed.`,
      );
    }

    const formattedValue = moment(fieldValue);

    // Support for plaintext display
    if (meta.plaintext) {
      const displayValue = this.getDisplayValue(formattedValue);

      return (
        <FieldLine {...this.props}>
          <StrapInput {...field} value={displayValue} plaintext />
        </FieldLine>
      );
    }

    const inputProps = {
      disabled: field.disabled,
    };

    return (
      <FieldLine {...this.props}>
        <Datetime
          value={formattedValue}
          onChange={this.handleChange}
          onBlur={this.handleBlur}
          dateFormat={dateFormat}
          timeFormat={timeFormat}
          inputProps={inputProps}
          closeOnSelect
        />
      </FieldLine>
    );
  }
}

export const DatePicker = withField(BaseDatePicker);
