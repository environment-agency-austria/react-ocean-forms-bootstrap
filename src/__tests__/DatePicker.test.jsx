import React from 'react';
import Datetime from 'react-datetime';
import { shallow } from 'enzyme';

import createMockFormatter from '../test-utils/createMockFormatter';
import DatePicker from '../DatePicker';
import FieldLine from '../FieldLine';

describe('<DatePicker />', () => {
  const FIELD_ID = 'field0';
  const FIELD_NAME = 'field0';
  const FIELD_LABEL = 'field0';

  const meta = {
    valid: true,
    error: undefined,
    isValidating: undefined,
    stringFormatter: createMockFormatter(),
    plaintext: false,
  };
  const field = {
    value: '',
    invalid: false,
    id: FIELD_ID,
    name: FIELD_NAME,
    disabled: false,
    onChange: jest.fn(),
    onBlur: jest.fn(),
  };

  const setup = props => shallow((
    <DatePicker
      label={FIELD_LABEL}
      meta={meta}
      field={field}
      {...props}
    />
  ));

  afterEach(() => {
    field.onChange.mockClear();
    field.onBlur.mockClear();
  });

  it('should render correctly', () => {
    const wrapper = setup();
    expect(wrapper).toMatchSnapshot();
  });

  it('should render a field line and a bootstrap input', () => {
    const wrapper = setup();

    expect(wrapper.find(FieldLine).exists()).toBeTruthy();
    expect(wrapper.find(Datetime).exists()).toBeTruthy();
  });

  it('should react on meta.plaintext correctly', () => {
    meta.plaintext = true;
    const wrapper = setup();
    expect(wrapper).toMatchSnapshot();
  });
});
