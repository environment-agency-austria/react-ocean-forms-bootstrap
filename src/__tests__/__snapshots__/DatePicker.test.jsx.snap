// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DatePicker /> should react on meta.plaintext correctly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <DatePicker
    dateFormat={null}
    field={
      Object {
        "disabled": false,
        "id": "field0",
        "invalid": false,
        "name": "field0",
        "onBlur": [MockFunction],
        "onChange": [MockFunction],
        "value": "",
      }
    }
    label="field0"
    meta={
      Object {
        "error": undefined,
        "isValidating": undefined,
        "plaintext": true,
        "stringFormatter": [MockFunction],
        "valid": true,
      }
    }
    placeholder="ojs_select_placeholder"
    timeFormat={null}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "children": <Input
        disabled={false}
        id="field0"
        invalid={false}
        name="field0"
        onBlur={[MockFunction]}
        onChange={[MockFunction]}
        plaintext={true}
        type="text"
        value=""
      >
        Invalid date
      </Input>,
      "dateFormat": null,
      "field": Object {
        "disabled": false,
        "id": "field0",
        "invalid": false,
        "name": "field0",
        "onBlur": [MockFunction],
        "onChange": [MockFunction],
        "value": "",
      },
      "info": undefined,
      "label": "field0",
      "meta": Object {
        "error": undefined,
        "isValidating": undefined,
        "plaintext": true,
        "stringFormatter": [MockFunction],
        "valid": true,
      },
      "placeholder": "ojs_select_placeholder",
      "prefix": undefined,
      "suffix": undefined,
      "timeFormat": null,
      "validators": undefined,
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": "Invalid date",
        "disabled": false,
        "id": "field0",
        "invalid": false,
        "name": "field0",
        "onBlur": [MockFunction],
        "onChange": [MockFunction],
        "plaintext": true,
        "type": "text",
        "value": "",
      },
      "ref": null,
      "rendered": "Invalid date",
      "type": [Function],
    },
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": <Input
          disabled={false}
          id="field0"
          invalid={false}
          name="field0"
          onBlur={[MockFunction]}
          onChange={[MockFunction]}
          plaintext={true}
          type="text"
          value=""
        >
          Invalid date
        </Input>,
        "dateFormat": null,
        "field": Object {
          "disabled": false,
          "id": "field0",
          "invalid": false,
          "name": "field0",
          "onBlur": [MockFunction],
          "onChange": [MockFunction],
          "value": "",
        },
        "info": undefined,
        "label": "field0",
        "meta": Object {
          "error": undefined,
          "isValidating": undefined,
          "plaintext": true,
          "stringFormatter": [MockFunction],
          "valid": true,
        },
        "placeholder": "ojs_select_placeholder",
        "prefix": undefined,
        "suffix": undefined,
        "timeFormat": null,
        "validators": undefined,
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": "Invalid date",
          "disabled": false,
          "id": "field0",
          "invalid": false,
          "name": "field0",
          "onBlur": [MockFunction],
          "onChange": [MockFunction],
          "plaintext": true,
          "type": "text",
          "value": "",
        },
        "ref": null,
        "rendered": "Invalid date",
        "type": [Function],
      },
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getSnapshotBeforeUpdate": true,
        },
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;

exports[`<DatePicker /> should render correctly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <DatePicker
    dateFormat={null}
    field={
      Object {
        "disabled": false,
        "id": "field0",
        "invalid": false,
        "name": "field0",
        "onBlur": [MockFunction],
        "onChange": [MockFunction],
        "value": "",
      }
    }
    label="field0"
    meta={
      Object {
        "error": undefined,
        "isValidating": undefined,
        "plaintext": false,
        "stringFormatter": [MockFunction],
        "valid": true,
      }
    }
    placeholder="ojs_select_placeholder"
    timeFormat={null}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "children": <Unknown
        className=""
        closeOnSelect={true}
        closeOnTab={true}
        dateFormat={null}
        defaultValue=""
        id="field0"
        input={true}
        inputProps={
          Object {
            "disabled": false,
          }
        }
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        onViewModeChange={[Function]}
        placeholder="ojs_select_placeholder"
        strictParsing={true}
        timeConstraints={Object {}}
        timeFormat={null}
        utc={false}
        value={null}
      />,
      "dateFormat": null,
      "field": Object {
        "disabled": false,
        "id": "field0",
        "invalid": false,
        "name": "field0",
        "onBlur": [MockFunction],
        "onChange": [MockFunction],
        "value": "",
      },
      "info": undefined,
      "label": "field0",
      "meta": Object {
        "error": undefined,
        "isValidating": undefined,
        "plaintext": false,
        "stringFormatter": [MockFunction],
        "valid": true,
      },
      "placeholder": "ojs_select_placeholder",
      "prefix": undefined,
      "suffix": undefined,
      "timeFormat": null,
      "validators": undefined,
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "className": "",
        "closeOnSelect": true,
        "closeOnTab": true,
        "dateFormat": null,
        "defaultValue": "",
        "id": "field0",
        "input": true,
        "inputProps": Object {
          "disabled": false,
        },
        "onBlur": [Function],
        "onChange": [Function],
        "onFocus": [Function],
        "onViewModeChange": [Function],
        "placeholder": "ojs_select_placeholder",
        "strictParsing": true,
        "timeConstraints": Object {},
        "timeFormat": null,
        "utc": false,
        "value": null,
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": <Unknown
          className=""
          closeOnSelect={true}
          closeOnTab={true}
          dateFormat={null}
          defaultValue=""
          id="field0"
          input={true}
          inputProps={
            Object {
              "disabled": false,
            }
          }
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          onViewModeChange={[Function]}
          placeholder="ojs_select_placeholder"
          strictParsing={true}
          timeConstraints={Object {}}
          timeFormat={null}
          utc={false}
          value={null}
        />,
        "dateFormat": null,
        "field": Object {
          "disabled": false,
          "id": "field0",
          "invalid": false,
          "name": "field0",
          "onBlur": [MockFunction],
          "onChange": [MockFunction],
          "value": "",
        },
        "info": undefined,
        "label": "field0",
        "meta": Object {
          "error": undefined,
          "isValidating": undefined,
          "plaintext": false,
          "stringFormatter": [MockFunction],
          "valid": true,
        },
        "placeholder": "ojs_select_placeholder",
        "prefix": undefined,
        "suffix": undefined,
        "timeFormat": null,
        "validators": undefined,
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "className": "",
          "closeOnSelect": true,
          "closeOnTab": true,
          "dateFormat": null,
          "defaultValue": "",
          "id": "field0",
          "input": true,
          "inputProps": Object {
            "disabled": false,
          },
          "onBlur": [Function],
          "onChange": [Function],
          "onFocus": [Function],
          "onViewModeChange": [Function],
          "placeholder": "ojs_select_placeholder",
          "strictParsing": true,
          "timeConstraints": Object {},
          "timeFormat": null,
          "utc": false,
          "value": null,
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getSnapshotBeforeUpdate": true,
        },
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;
